riscv/proc_trace.cc

```diff
@@ -0,0 +1,325 @@
+            return;
+        }
+
+        // Limit value to the size given
+        if (size != 8)
+            res &= ((uint64_t(1) << (size * 8)) - 1);
+
+        assert(m_is_load);
+        assert(m_dtype == DTYPE_LOAD);
+        record_data(addr, (uint8_t*)&res, size);
+    }
+}
+
+void proc_trace_t::record_store(reg_t addr, uint64_t val, size_t size, data_src_t src) {
+    if (m_dtrace_fd) {
+        if (!m_has_insn || m_is_amo)
+            return;
+
+        // Limit value to the size given
+        if (size != 8)
+            val &= ((uint64_t(1) << (size * 8)) - 1);
+
+        assert(m_is_store);
+        assert(m_dtype == DTYPE_STORE);
+        record_data(addr, (uint8_t*)&val, size);
+    }
+}
+
+/* Atomic and CSR accesses have either both load and store data, or store data and an operand. For
+   CSRs and unified atomics, both values are reported via data, with the store data in the LSBs and
+   the load data or operand in the MSBs. */
+void proc_trace_t::record_amo(reg_t addr, uint64_t load_data, uint64_t store_data, size_t size) {
+    if (m_dtrace_fd) {
+        reg_t insn_31_27 = (m_insn_binary >> 27) & ((insn_bits_t(1) << 5) - 1);
+        switch (insn_31_27) {
+        case 0: set_dtype(DTYPE_ATOMIC_ADD); break;
+        case 1: set_dtype(DTYPE_ATOMIC_SWAP); break;
+        case 4: set_dtype(DTYPE_ATOMIC_XOR); break;
+        case 8: set_dtype(DTYPE_ATOMIC_OR); break;
+        case 12: set_dtype(DTYPE_ATOMIC_AND); break;
+        case 16: case 24: set_dtype(DTYPE_ATOMIC_MIN); break;
+        case 20: case 28: set_dtype(DTYPE_ATOMIC_MAX); break;
+        default: assert(0); break;
+        }
+
+        uint64_t pair[2] = {load_data, store_data};
+        record_data(addr, (uint8_t*)pair, 2 * size);
+    }
+}
+
+void proc_trace_t::record_csr_get_trace(int which, insn_t insn, bool write, reg_t load_data, size_t size) {
+    // Record the current csr being processed as there may be a set to go with this
+    m_csr_which = which;
+    m_csr_write = write;
+    m_csr_load_data = load_data;
+    m_csr_size = size;
+
+    reg_t func3 = insn.rm();
+    switch (func3) {
+    case 1: // CSRRW
+    case 5: // CSRRWI
+        set_dtype(DTYPE_CSR_READ_WRITE);
+        break;
+    case 2: // CSRRS
+    case 6: // CSRRSI
+        set_dtype(DTYPE_CSR_READ_SET);
+        break;
+    case 3: // CSRRC
+    case 7: // CSRRCI
+        set_dtype(DTYPE_CSR_READ_CLEAR);
+        break;
+    default:
+        break;
+    }
+
+    // If not being written then the trace must be written here
+    if (!m_csr_write) {
+        assert(m_csr_which != CSR_INVALID);
+        uint64_t pair[2] = {load_data, 0};
+        record_data(m_csr_which, (uint8_t*)pair, 2 * size);
+    }
+}
+
+void proc_trace_t::record_csr_set_trace(int which, reg_t store_data) {
+    // mret is a special case where there is no call to csr_get to go with this.
+    // In this case the csr must be MSTATUS and it does not get recorded.
```

Comment:
Why doesn't it get recorded? Is that specified in the trace spec?

Reply:
I have removed the magic number in this function and replaced with a constant defined in encoding.h